File: 2025/05/main.py
def parse():
    fresh_ranges,to_check = open("input.txt").read().split("\n\n")
    ranges = [tuple(map(int, line.split("-"))) for line in fresh_ranges.splitlines()]
    ids = [int(line) for line in to_check.splitlines()]
    return ranges, ids

def is_fresh(num,ranges):
    for low,high in ranges:
        if low <= num <= high:
            return True
    return False

def part1(ranges,ids):
    return sum(1 for num in ids if is_fresh(num,ranges))

def part2(ranges):
    sorted_ranges = sorted(ranges)
    merged = [sorted_ranges[0]]
    for low,high in sorted_ranges[1:]:
        last_low,last_high = merged[-1]
        if low <= last_high + 1:
            merged[-1] = (last_low, max(high, last_high))
        else:
            merged.append((low, high))
    return sum(high-low+1 for low,high in merged)

if __name__ == "__main__":
    ranges,ids = parse()
    print("Part 1:",part1(ranges,ids))
    print("Part 2:",part2(ranges))

File: 2025/05/main.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Range struct {
	low, high int
}

func parse() ([]Range, []int) {
	data, _ := os.ReadFile("input.txt")
	parts := strings.Split(string(data), "\n\n")

	var ranges []Range
	for _, line := range strings.Split(strings.TrimSpace(parts[0]), "\n") {
		var r Range
		fmt.Sscanf(line, "%d-%d", &r.low, &r.high)
		ranges = append(ranges, r)
	}

	var ids []int
	for _, line := range strings.Split(strings.TrimSpace(parts[1]), "\n") {
		id, _ := strconv.Atoi(line)
		ids = append(ids, id)
	}

	return ranges, ids
}

func isFresh(num int, ranges []Range) bool {
	for _, r := range ranges {
		if r.low <= num && num <= r.high {
			return true
		}
	}
	return false
}

func part1(ranges []Range, ids []int) int {
	count := 0
	for _, id := range ids {
		if isFresh(id, ranges) {
			count++
		}
	}
	return count
}

func part2(ranges []Range) int {
	sorted := make([]Range, len(ranges))
	copy(sorted, ranges)
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i].low < sorted[j].low
	})

	merged := []Range{sorted[0]}

	for _, r := range sorted[1:] {
		last := &merged[len(merged)-1]
		if r.low <= last.high+1 {
			last.high = max(last.high, r.high)
		} else {
			merged = append(merged, r)
		}
	}

	total := 0
	for _, r := range merged {
		total += r.high - r.low + 1
	}
	return total
}

func main() {
	ranges, ids := parse()
	fmt.Println("Part 1:", part1(ranges, ids))
	fmt.Println("Part 2:", part2(ranges))
}
